### PR TITLE
sstables/sstable: validate_checksums(): force-check EOF

### PR DESCRIPTION
**Pull request by denesb**
EOF is only guarateed to be set if one tried to read past the end of the file. So when checking for EOF, also try to read one more byte. This should force the EOF flag into a correct value. We can then check that the read yielded 0 bytes and that EOF is set. Checking both is redundant but is more explicit about the intent.
This should ensure that `validate_checksums()` will not falsely declare the validation to have failed.

Fixes: https://github.com/scylladb/scylladb/issues/11190

### CODE REVIEW

File: sstables/sstables.cc [hunk 1]
```diff
@@ -2518,9 +2518,12 @@ static future<bool> do_validate_uncompressed(input_stream<char>& stream, const c
         offset += buf.size();
     }
 
-    if (!stream.eof()) {
-        sstlog.error("Chunk count mismatch between CRC.db and Data.db at offset {}: expected {} chunks but data file has more", offset, checksum.checksums.size());
-        valid = false;
+    {
+        auto buf = co_await stream.read_exactly(4 * 1024);
+        if (!buf.empty() || !stream.eof()) {
```
Comment (nyh) [inline sstables/sstables.cc:2523]: nitpick: no need to check eof() - read_exactly returning an empty buffer already means exactly what you want, that there were no more bytes to read.

By the way, you can also read 1 byte, no need to specifically read 4 * 1024 bytes - readers might think this number has some sort of significance.

**Review by nyh — APPROVED** [review]
Looks good.
By the way, I think you were too hard on eof() in the commit message. I think it works exactly as it should: Like the Unix feof() and similar eof markers in all sorts of languages and libraries, eof() becomes true only after you read until the end-of-file once in the past. For example, if you open an empty file and not read anything yet, feof() will be false until you read from the file.
We should have a comment in Seastar's iostream.hh that says this explicitly... (I checked - we don't).

**Comment by avikivity** [comment]
> Looks good. By the way, I think you were too hard on eof() in the commit message. I think it works exactly as it should: Like the Unix feof() and similar eof markers in all sorts of languages and libraries, eof() becomes true only after you read until the end-of-file once in the past. For example, if you open an empty file and not read anything yet, feof() will be false until you read from the file. We should have a comment in Seastar's iostream.hh that says this explicitly... (I checked - we don't).

Agree.

**Comment by denesb** [comment]
new in v2: don't blame EOF, instead acknowledge that the code did not check it properly before; also add a comment to the check

File: sstables/sstables.cc (hunk 1 above)
Comment (nyh) [reply sstables/sstables.cc:2526]: Oh, and you can also do read() instead of read_exactly(N) with any specific N.

**Comment by denesb** [comment]
new in v3: use `read()` instead of `read_exactly()`.

**Comment by nyh** [comment]
> new in v3: use `read()` instead of `read_exactly()`.

Good. But in this case I think it's even more obvious that the check for empty() is enough and the eof() check isn't needed (it doesn't make the code wrong - but it's just weird).

**Comment by denesb** [comment]
new in v4: remove redundant eof check

**Review by avikivity — APPROVED** [review]

**Pull request closed without merging**